### DISCUSSION TITLE
Does the keyboard remapping use xkb instead of xmodmap for xorg?

### DISCUSSION THREAD

**Discussion by Srijan1214** [sezanzeb/input-remapper#52] [Q&A]
Hi,

I am very new with the whole keyboard remappings in Linux.

I really like this application. I use Ubuntu 20.04 with Gnome 3.36. I had a lot of issues with `xmodmap` (which was recommended on most of my online searches) only to find out that `xmodmap` is no longer supported by Gnome.

I just wanted to ask if this application uses `xkb` under the hood for `xorg` environments. Or does it do something completely different?

Thank you for your time.

**Comment by sezanzeb** [comment]
It works with evdev

1. It grabs a device (e.g. /dev/input/event3), so that the key events won't reach X11/Wayland anymore [source](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/injector.py#L182)
2. Reads the events from it (`evtest` can do it, you can also do `cat /dev/input/event3` which yields binary stuff) [source](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/injector.py#L413)
3. Looks up the mapping if that event maps to anything [source](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/keycode_mapper.py#L421)
4. Injects the output event in a new device that key-mapper created (another new path in /dev/input, device name is suffixed by "mapped") [source](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/keycode_mapper.py#L227), [new device](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/injector.py#L324)
5. Forwards any events that should not be mapped to anything in another new device (device name is suffixed by "forwarded") [source](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/keycode_mapper.py#L232), [new device](https://github.com/sezanzeb/key-mapper/blob/0.7.0/keymapper/injection/injector.py#L342)

This stuff is going on as a daemon in the background, also see https://github.com/sezanzeb/key-mapper/blob/main/readme/development.md#files

Additional optional stuff can be done with xkb on top of the key-mapper devices ("mapped" and "forwarded") https://github.com/sezanzeb/key-mapper/issues/33, which is currently in development
